File: services/edge-stub/tests/contract.py
#!/usr/bin/env python3
"""Contract check: openapi.yaml paths section defines routes the edge stub implements."""
from __future__ import annotations

import sys
from pathlib import Path

try:
    import yaml
except ImportError:
    print("contract: install PyYAML (pip install pyyaml)", file=sys.stderr)
    sys.exit(1)

ROOT = Path(__file__).resolve().parents[3]
SPEC_PATH = ROOT / "docs" / "api" / "openapi.yaml"

REQUIRED_PATHS = ["/api/v1/chat", "/api/v1/generate", "/api/v1/embed", "/api/v1/models"]
RECOMMENDED_PATHS = ["/health"]


def main() -> int:
    try:
        spec = yaml.safe_load(SPEC_PATH.read_text(encoding="utf-8"))
    except OSError as exc:
        print(f"contract: could not read spec at {SPEC_PATH}: {exc}", file=sys.stderr)
        return 1

    paths = set((spec or {}).get("paths") or {})
    missing_required = [p for p in REQUIRED_PATHS if p not in paths]
    if missing_required:
        print(
            "contract: spec paths section is missing required paths:",
            ", ".join(missing_required),
            file=sys.stderr,
        )
        return 1

    missing_recommended = [p for p in RECOMMENDED_PATHS if p not in paths]
    if missing_recommended:
        print(
            "contract: spec should also define (recommended):",
            ", ".join(missing_recommended),
        )

    print("contract: spec contains required paths")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: services/edge-stub/tests/contract.mjs
/**
 * Contract check: local OpenAPI spec defines paths the edge stub implements.
 * Run from repo root: node services/edge-stub/tests/contract.mjs
 */
import { readFileSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "../../..");
const SPEC_PATH = join(ROOT, "docs/api/openapi.yaml");

const REQUIRED_PATHS = ["/api/v1/chat", "/api/v1/generate", "/api/v1/embed", "/api/v1/models"];
const RECOMMENDED_PATHS = ["/health"];

function specContainsPath(specText, path) {
  if (specText.includes(path)) return true;
  const pathKey = path.startsWith("/") ? path : "/" + path;
  return specText.includes(pathKey + ":") || specText.includes(" " + pathKey);
}

let text;
try {
  text = readFileSync(SPEC_PATH, "utf8");
} catch (e) {
  console.error("contract: could not read spec at", SPEC_PATH, e.message);
  process.exit(1);
}

const missingRequired = REQUIRED_PATHS.filter((p) => !specContainsPath(text, p));
if (missingRequired.length) {
  console.error("contract: spec is missing required paths:", missingRequired.join(", "));
  process.exit(1);
}

const missingRecommended = RECOMMENDED_PATHS.filter((p) => !specContainsPath(text, p));
if (missingRecommended.length) {
  console.warn("contract: spec should also define (recommended):", missingRecommended.join(", "));
}

console.log("contract: spec contains required paths");
process.exit(0);
